### ISSUE TITLE
Considerations about multiple Ycmd instances

### ISSUE THREAD

**Issue by iogf** [ycm-core/ycmd#1327]
Is it safe to have just a single ycmd instance for all of a set of IDE instances?
Let us suppose there are two IDE instances which are opened with files from two projects.  Let us say a javascript and a c++ one. Would ycmd work safely to give completions for both projects if there were just one ycmd instance being used by the two IDE applications?

**Comment by bstaletic** [comment]
Shared sessions are tricky and ycmd has never made any efforts to make them work. There are a few reasons for that:

- Seemingly very low, even if not zero, demand.
- The number of core maintainers behind ycmd.
- Finally, some semantic completion engines assume that sessions aren't shared.

So a shared instance for the same project and the same set of files is just out of the question. On the other hand, you're talking about separate projects and even separate completers that only share the same (for lack of a better word) core ycmd functionality. Since two clients would be listening to ip:port, HMAC would be responsible for discerning which messages belong to client X and which to client Y. To me, it looks like what you describe would work as long as there's no overlap between clients, however, since that constraint is not enforceable, we can't provide any guarantees. Nor would we consider bugs that arise from shared sessions, for the reason number 2 in the above list.

TL;DR: ¯\\\_(ツ)\_/¯ Maybe it can work, maybe it can't. If it breaks, keep the pieces.

**Comment by iogf** [comment]
I see.
